aggregator/prover/prover.go

```diff
@@ -2,81 +2,254 @@ package prover
+	stream                                     pb.AggregatorService_ChannelServer
 }
 
-// NewProver creates a new Prover
-func NewProver(proverURI string) *Prover {
-	const checkWaitInSeconds = 20
-	ctx := context.Background()
-	tickerCheckConnection := time.NewTicker(checkWaitInSeconds * time.Second)
-	opts := []grpc.DialOption{
-		// TODO: once we have user and password for prover server, change this
-		grpc.WithTransportCredentials(insecure.NewCredentials()),
+// New returns a new Prover instance.
+func New(stream pb.AggregatorService_ChannelServer, intervalFrequencyToGetProofGenerationState types.Duration) (*Prover, error) {
+	p := &Prover{
+		stream: stream,
+		IntervalFrequencyToGetProofGenerationState: intervalFrequencyToGetProofGenerationState,
 	}
-	proverConn, err := grpc.Dial(proverURI, opts...)
+	status, err := p.Status()
 	if err != nil {
-		log.Fatalf("fail to dial: %v", err)
+		return nil, err
 	}
+	p.id = status.ProverId
+	return p, nil
+}
 
-	proverClient := proverclientpb.NewZKProverServiceClient(proverConn)
-	prover := &Prover{URI: proverURI, Client: proverClient, Conn: proverConn}
+// ID returns the Prover ID.
+func (p *Prover) ID() string { return p.id }
 
-	go func() {
-		waitTick(ctx, tickerCheckConnection)
-		for {
-			prover.checkConnection(ctx, tickerCheckConnection)
-		}
-	}()
+// Status gets the prover status.
+func (p *Prover) Status() (*pb.GetStatusResponse, error) {
+	req := &pb.AggregatorMessage{
+		Request: &pb.AggregatorMessage_GetStatusRequest{
+			GetStatusRequest: &pb.GetStatusRequest{},
+		},
+	}
+	res, err := p.call(req)
+	if err != nil {
+		return nil, err
+	}
+	if msg, ok := res.Response.(*pb.ProverMessage_GetStatusResponse); ok {
+		return msg.GetStatusResponse, nil
+	}
+	return nil, errors.New("Bad response") // FIXME(pg)
+}
 
-	return prover
+// IsIdle returns true if the prover is idling.
+func (p *Prover) IsIdle() bool {
+	status, err := p.Status()
+	if err != nil {
+		log.Warnf("Error asking status for prover ID %s: %w", p.ID(), err)
+		return false
+	}
+	return status.Status == pb.GetStatusResponse_IDLE
 }
 
-func (p *Prover) checkConnection(ctx context.Context, ticker *time.Ticker) {
-	state := p.Conn.GetState()
-	log.Debugf("Checking connection to prover %v. State: %v", p.URI, state)
+// BatchProof instructs the prover to generate a batch proof for the provided
+// input. It returns the ID of the proof being computed.
+func (p *Prover) BatchProof(input *pb.InputProver) (string, error) {
+	req := &pb.AggregatorMessage{
+		Request: &pb.AggregatorMessage_GenBatchProofRequest{
+			GenBatchProofRequest: &pb.GenBatchProofRequest{Input: input},
+		},
+	}
+	res, err := p.call(req)
+	if err != nil {
+		return "", err
+	}
+	if msg, ok := res.Response.(*pb.ProverMessage_GenBatchProofResponse); ok {
+		switch msg.GenBatchProofResponse.Result {
+		case pb.Result_OK:
+			return msg.GenBatchProofResponse.Id, nil
+		case pb.Result_ERROR:
+			return "", errors.New("GenBatchProofResponse.Result: ERROR")
```

Comment:
This error could be defined as constants so they could be checked properly from outside.

Reply:
@ToniRamirezM good point, the errors were still not properly handled in this file.
I tried to handle them better with 59e28c843ca4745b69e6c51869958d30c2b87780 and I added error types as per your suggestion in 5a3fe36d925d00ec4450ed172c97c032a66af703